### PR TITLE
Introduce Slotty stuff

### PR DESCRIPTION
**Pull request by addshore**
Just an idea now to discuss

### CODE REVIEW

File: src/Slotty.php [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+ * @since 1.9.0
+ */
+interface Slotty {
+
+	/**
+	 * @return array
+	 *     Keys: The name of the slot.
+	 *     Values: The value of the slot.
+	 *         array[]|int[]|string[]|bool[]|float[]
+	 *         A possibly nested structure consisting of only arrays and scalar values
```
Comment (Benestar) [inline src/Slotty.php]: I thought those would be DataValue objects?

File: src/Serializers/SlottySerializer.php [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+use Serializers\DispatchableSerializer;
+use Serializers\Exceptions\SerializationException;
+use Wikibase\DataModel\Slotty;
+
+/**
+ * Package private
+ *
+ * @licence GNU GPL v2+
+ * @author Adam Shorland
+ * @since 1.9.0
```
Comment (JeroenDeDauw) [inline src/Serializers/SlottySerializer.php]: wow, 3 different since tags. I guess this supports our mission of diversity!

File: src/Slotty.php (hunk 1 above)
Comment (addshore) [inline src/Slotty.php]: Well, I guess they could be, but I see no reason they couldn't just be strings.
If we had them as DataValues and then pushed to use this in the API for urls of sitelinks and datatypes or snaks etc that would require a breaking change in the API or more hacks.

File: src/Serializers/SlottySerializer.php [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+ */
+class SlottySerializer implements DispatchableSerializer {
```
Comment (Benestar) [inline src/Serializers/SlottySerializer.php]: I wonder why we need a serializer for this if the object's "slots" are just returned as they exist. If serialization of slots should be that simple, just get rid of that serializer and use `$object->getSlots()` directly.

**Comment by Benestar** [comment]
This way we need to introduce a subclass of `SiteLink` called sth like `SlottySiteLink` that implements the interface and a thing for snaks as well, also for terms. These objects should then become part of this serialization component.

File: src/Serializers/SlottySerializer.php [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+	/**
+	 * @since 1.0
+	 *
+	 * @param Slotty $object
+	 *
+	 * @return array|int|string|bool|float A possibly nested structure consisting of only arrays and scalar values
+	 * @throws SerializationException
+	 */
+	public function serialize( $object ) {
+		return $object->getSlots();
```
Comment (thiemowmde) [inline src/Serializers/SlottySerializer.php]: I do have two major problems with this proposal:

1. The name is [killing me](http://dict.leo.org/?search=slutty).
2. What's the point of having support for "slots" that can contain nothing but scalar values? What this basically means is: All code working with these "slots" must do it's own (de)serialization to make sure getSlots only returns values that are already in the correct intermediate array format for the JSON and XML serialization. The knowledge about how to (de)serialize a specific "slot" will be spread across different components. If you do this the other way around and have one or many actual "SlotSerializers" that know how to (de)serialize all supported "slots", what's the difference and advantage compared to an actual change to the DataModel and (de)serializers then?

TL;DR: The effect of this proposal is that (de)serializer code will not be in a (de)serializer component any more, but copy-pasted all over the code base where one of the fields must be (de)serialized.

Comment (JeroenDeDauw) [inline src/Serializers/SlottySerializer.php]: I have similar concerns already written down on another PR that was doing the same or something very similar. Can't find it now though

**Comment by JeroenDeDauw** [comment]
Assuming this can be closed. Reopen if not
